### PR TITLE
Rashuai/gathernd op

### PR DESCRIPTION
**Pull request by RandyShuai**
Define gather_np as contrib op for cpu.

### CODE REVIEW

File: onnxruntime/contrib_ops/cpu/gather_nd.cc [hunk 1]
```diff
@@ -110,8 +110,8 @@ Status GatherNDString<Tind>::Compute(OpKernelContext* context) const {
   Prepare p;
   ONNXRUNTIME_RETURN_IF_ERROR(PrepareForCompute<Tind>(context, p));
 #pragma omp parallel for
-  for (size_t i = 0; i < p.element_offsets.size(); ++i) {
-    for (size_t j = 0; j < p.element_to_copy; ++j) {
+  for (int64_t i = 0; i < static_cast<int64_t>(p.element_offsets.size()); ++i) {
```
Comment (snnn) [inline onnxruntime/contrib_ops/cpu/gather_nd.cc]: Does it work on 32 bits systems?

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc [hunk 2]
```diff
@@ -498,6 +498,71 @@ The bounding box coordinates corresponding to the selected indices can then be o
...
+            ->add_dim() = data_shape.dim(i);
+       }
+    })
+    .SetDoc(R"DOC(
+Given `data` tensor of rank r >= 1, and `indices` tensor of rank q >= 1, gather
+slices of `data` into an output tensor of rank q - 1 + r - indices[-1].
+Example 1:
+  data    = [[0,1],[2,3]]
+  indices = [[0,0],[1,1]]
+  output  = [0,3]
```
Comment (HectorSVC) [inline onnxruntime/core/graph/contrib_ops/contrib_defs.cc]: >[0,3] [](start = 12, length = 5)

for this case, shouldn't the output be [[0],[3]]?

Comment (RandyShuai) [reply onnxruntime/core/graph/contrib_ops/contrib_defs.cc]: Thanks for the comment!
But it should be [0,3]:
[0,1] indexed to scaler 0,
[2,3] indexed to scaler 3,
Together with outer bracket of indices tensor we got [0,3]

File: onnxruntime/contrib_ops/cpu/gather_nd.cc (hunk 1 above)
Comment (RandyShuai) [reply onnxruntime/contrib_ops/cpu/gather_nd.cc]: Good question!
On 32 bit os - size() has 4 bytes while on 64 it is 8, right?
Either way, conversion to int64_t is valid by cpp standard.
The only loophole here may be that when size() has 8 bytes and it is big enough, we may receive a wrong int64_t after cast, however, the situation should only exist theoretically...

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc (hunk 2 above)
Comment (RandyShuai) [reply onnxruntime/core/graph/contrib_ops/contrib_defs.cc]: BTW, to get [[0],[3]], indices need to be [[[0,0]],[[1,1]]],

File: onnxruntime/contrib_ops/cpu/gather_nd.cc (hunk 1 above)
Comment (RandyShuai) [reply onnxruntime/contrib_ops/cpu/gather_nd.cc]: FYI, https://en.cppreference.com/w/cpp/language/implicit_conversion

Comment (snnn) [reply onnxruntime/contrib_ops/cpu/gather_nd.cc]: Is it ok to use int64_t as loop index on 32 bits system, for openmp?

Comment (RandyShuai) [reply onnxruntime/contrib_ops/cpu/gather_nd.cc]: Should be fine as long as it is supported cpp primitive integer type - believe it is all compile's concern, not coder.

File: onnxruntime/contrib_ops/cpu/gather_nd.h [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+               element_to_copy (0),
+               element_offsets (0) {}
+  }; // struct Prepare
+
+  template<typename Tind>
+  Status PrepareForCompute(OpKernelContext* context, Prepare& p) const;
+}; // class GatherNDBase
+
+template<typename Tind>
+class GatherNDString final : public OpKernel, protected GatherNDBase {
```
Comment (HectorSVC) [inline onnxruntime/contrib_ops/cpu/gather_nd.h]: >GatherNDString [](start = 6, length = 14)

no need to have 2 separate class, we can get the input data type.
[https://github.com/Microsoft/onnxruntime/blob/master/onnxruntime/core/providers/cpu/tensor/gather.cc#L77](https://github.com/Microsoft/onnxruntime/blob/master/onnxruntime/core/providers/cpu/tensor/gather.cc#L77)

Comment (RandyShuai) [reply onnxruntime/contrib_ops/cpu/gather_nd.h]: Yes I see that from Gather implementation.
In Gather, the implementation keeps a BOOL - "is_string_type", then in a for-for loop, it judges every single time whether the input type is a string.
By template, we could move this efforts to compile time.

File: onnxruntime/contrib_ops/cpu/gather_nd.h [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+               element_bytes   (0),
+               element_to_copy (0),
+               element_offsets (0) {}
+  }; // struct Prepare
+
+  template<typename Tind>
+  Status PrepareForCompute(OpKernelContext* context, Prepare& p) const;
+}; // class GatherNDBase
+
+template<typename Tind>
```
[the remark this answers is not in the document]
Comment (RandyShuai) [reply onnxruntime/contrib_ops/cpu/gather_nd.h]: Thanks for the heads up!
However, as explained in above comment, the template class for string type is there for a reason - we do not want to have extra computing in run time whenever it is possible.

File: onnxruntime/contrib_ops/cpu/gather_nd.h (hunk 3 above)
Comment (HectorSVC) [reply onnxruntime/contrib_ops/cpu/gather_nd.h]: the reason we want to get rid of template is to keep the binary size as small as possible. I think it won't impact much put the if inside the loop.

---
In reply to: [241870901](https://github.com/Microsoft/onnxruntime/pull/170#discussion_r241870901) [](ancestors = 241870901)

Comment (RandyShuai) [reply onnxruntime/contrib_ops/cpu/gather_nd.h]: Got it, now they are removed.

**Pull request merged**